### PR TITLE
fix: provide unique `fix` and `fix.range` objects in lint messages

### LINKED ISSUE #16716: Bug: suggestions are broken on processored code if `suggest` array contain more that 2 items
### Environment

Node version: 16.18.1
npm version: 
Local ESLint version: 8.30.0
Global ESLint version: 
Operating System: 

### What parser are you using?

Other

### What did you do?

`suggest` array contains 2 items

https://user-images.githubusercontent.com/7361780/209718712-417b52b5-5e16-402b-a133-c9a151a75d4f.mov

`suggest` array contains 1 item

https://user-images.githubusercontent.com/7361780/209718742-5104ae35-8d33-4634-a53f-f66ebef8cd27.mov

This behaviour only happens on processed code (in my case `*.js/*.ts` since I'm the maintainer of [`graphql-eslint`](https://github.com/B2o5T/graphql-eslint) package.
There is no this problem when I apply suggestions on `*.graphql` files directly

I verified another rule that does not do `fixer.replaceText` but `fixer.remove` and received exact same behaviour, suggestions are broken if I add 2nd item.

### What did you expect to happen?

should not break code if there are more than 2 items in suggest array on processed code

### What actually happened?

Suggestions works on `*.js` files only if there is 1 item in the `suggest` array
Suggestions works well on `*.graphql` files if there are 2 items in suggest

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by mdjermanovic**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

Fixes #16716

#### What changes did you make? (Give an overview)

Updated `report-translator` to always create new `fix` objects and `fix.range` arrays. This ensures that lint messages do not share same instances, which could be problematic for processors that mutate these objects.

Added unit tests for `report-translator`, and one higher-level `linter` test that specifically targets the case described in #16716.

### CODE REVIEW

**Comment by fasttime** [comment]
Thanks for this PR. I think that a processor `postprocess()` function can still return objects with non-unique references, so if problems are passed from a processor to the next one, there is still a chance that some of them will have identical `fix` and `fix.range`?

**Comment by mdjermanovic** [comment]
> I think that a processor `postprocess()` function can still return objects with non-unique references, so if problems are passed from a processor to the next one, there is still a chance that some of them will have identical `fix` and `fix.range`?

That looks like something that should rather be fixed in the processor? I think it's unlikely that "normal" processing of messages with unique references produces an output with shared references.

**Comment by fasttime** [comment]
LGTM!

**Review by fasttime — APPROVED** [review]

File: lib/linter/report-translator.js [hunk 1]
```diff
@@ -100,6 +100,22 @@ function normalizeReportLoc(descriptor) {
     return descriptor.node.loc;
 }
 
+/**
+ * Clones the given fix object.
+ * @param {Fix|null} fix The fix to clone.
+ * @returns {Fix|null} Deep cloned fix object or `null` if `null` was passed in.
```
Comment (nzakas) [inline lib/linter/report-translator.js:106]:
```suggestion
 * @returns {Fix|null} Deep cloned fix object or `null` if `null` or `undefined` was passed in.
```

**Review by nzakas — APPROVED** [review]
LGTM. Just left one suggestion to clarify JSDoc, but otherwise, ready to merge.

**Pull request merged**